pkg/deployer/helm/ensure.go

```diff
@@ -74,40 +86,63 @@ func (h *Helm) ApplyFiles(ctx context.Context, files map[string]string, exports
 			},
 		}
 	}
+	h.ProviderStatus.ManagedResources = managedResources
 
-	if h.ProviderStatus != nil {
-		if err := h.cleanupOrphanedResources(ctx, targetClient, h.ProviderStatus.ManagedResources, objects); err != nil {
-			err = fmt.Errorf("unable to cleanup orphaned resources: %w", err)
-			return lsv1alpha1helper.NewWrappedError(err,
-				currOp, "CleanupOrphanedResources", err.Error())
-		}
+	if err := h.cleanupOrphanedResources(ctx, targetClient, h.ProviderStatus.ManagedResources, objects); err != nil {
+		err = fmt.Errorf("unable to cleanup orphaned resources: %w", err)
+		return lsv1alpha1helper.NewWrappedError(err,
+			currOp, "CleanupOrphanedResources", err.Error())
 	}
 
-	status := &helmv1alpha1.ProviderStatus{
-		TypeMeta: metav1.TypeMeta{
-			APIVersion: helmv1alpha1.SchemeGroupVersion.String(),
-			Kind:       "ProviderStatus",
-		},
-		ManagedResources: managedResources,
-	}
-	statusData, err := encodeStatus(status)
+	h.DeployItem.Status.ProviderStatus, err = encodeStatus(h.ProviderStatus)
 	if err != nil {
 		return lsv1alpha1helper.NewWrappedError(err,
-			currOp, "UpdateProviderStatus", err.Error())
+			currOp, "ProviderStatus", err.Error())
+	}
+	if err := h.kubeClient.Status().Update(ctx, h.DeployItem); err != nil {
+		return lsv1alpha1helper.NewWrappedError(err,
+			currOp, "UpdateStatus", err.Error())
 	}
 
 	if err := h.createOrUpdateExport(ctx, exports); err != nil {
 		return lsv1alpha1helper.NewWrappedError(err,
 			currOp, "CreateExport", err.Error())
 	}
 
-	h.DeployItem.Status.Phase = lsv1alpha1.ExecutionPhaseSucceeded
-	h.DeployItem.Status.ProviderStatus = statusData
-	h.DeployItem.Status.ObservedGeneration = h.DeployItem.Generation
-	if err := h.kubeClient.Status().Update(ctx, h.DeployItem); err != nil {
+	if h.ProviderConfiguration.HealthChecks.DisableDefault {
+		h.DeployItem.Status.Phase = lsv1alpha1.ExecutionPhaseSucceeded
+		h.DeployItem.Status.ObservedGeneration = h.DeployItem.Generation
+		h.DeployItem.Status.LastError = nil
+		return nil
+	}
+
+	return h.CheckResourcesHealth(ctx, targetClient)
+}
+
+// CheckResourcesHealth checks if the managed resources are Ready/Healthy.
+func (h *Helm) CheckResourcesHealth(ctx context.Context, client client.Client) error {
+	var (
+		currOp = "CheckResourcesHealthHelm"
+	)
+
+	if len(h.ProviderStatus.ManagedResources) == 0 {
+		return nil
+	}
+
+	objects := make([]*unstructured.Unstructured, len(h.ProviderStatus.ManagedResources))
+	for i, ref := range h.ProviderStatus.ManagedResources {
+		obj := kutil.ObjectFromTypedObjectReference(&ref)
+		objects[i] = obj
+	}
+
+	timeout, _ := time.ParseDuration(h.ProviderConfiguration.HealthChecks.Timeout)
+	if err := health.WaitForObjectsHealthy(ctx, timeout, h.log, client, objects); err != nil {
 		return lsv1alpha1helper.NewWrappedError(err,
-			currOp, "UpdateStatus", err.Error())
+			currOp, "CheckResourcesReadiness", err.Error())
 	}
```

Comment:
if this func runs into a timeout, the health is reconciled in the next loop.
Which means that the timeout in the provider configuration would have no real effect.

Depending on the desired we should either check the timeout differently or introduce another unretryable error.
WDYT?

Reply:
Fixed with a new unrecoverable error type `ErrorHealthCheckTimeout`.